src/daft-plan/src/logical_optimization/rules/eliminate_cross_join.rs

```diff
@@ -0,0 +1,718 @@
+///
+/// A valid join means:
+/// 1. All referenced column of the left side is from the left schema, and
+///    all referenced column of the right side is from the right schema.
+/// 2. Or opposite. All referenced column of the left side is from the right schema,
+///    and the right side is from the left schema.
+///
+pub fn find_valid_equijoin_key_pair(
+    left_key: ExprRef,
+    right_key: ExprRef,
+    left_schema: SchemaRef,
+    right_schema: SchemaRef,
+) -> DaftResult<Option<(ExprRef, ExprRef)>> {
+    let left_using_columns = column_refs(left_key.clone());
+    let right_using_columns = column_refs(right_key.clone());
+
+    // Conditions like a = 10, will be added to non-equijoin.
+    if left_using_columns.is_empty() || right_using_columns.is_empty() {
+        return Ok(None);
+    }
+
+    if check_all_columns_from_schema(&left_using_columns, &left_schema)?
+        && check_all_columns_from_schema(&right_using_columns, &right_schema)?
+    {
+        return Ok(Some((left_key, right_key)));
+    } else if check_all_columns_from_schema(&right_using_columns, &left_schema)?
+        && check_all_columns_from_schema(&left_using_columns, &right_schema)?
+    {
+        return Ok(Some((right_key, left_key)));
+    }
+
+    Ok(None)
+}
+
+/// Return all references to columns in this expression.
+fn column_refs(expr: ExprRef) -> HashSet<Arc<str>> {
+    let mut set = HashSet::new();
+    expr.apply(|expr| {
+        if let Expr::Column(col) = expr.as_ref() {
+            set.insert(col.clone());
+        }
+        Ok(TreeNodeRecursion::Continue)
+    })
+    .expect("traversal is infallible");
+    set
+}
+
+/// Can this data type be used in hash join equal conditions??
+/// Data types here come from function 'equal_rows', if more data types are supported
+/// in equal_rows(hash join), add those data types here to generate join logical plan.
+pub fn can_hash(data_type: &DataType) -> bool {
+    match data_type {
+        DataType::Null => true,
+        DataType::Boolean => true,
+        DataType::Int8 => true,
+        DataType::Int16 => true,
+        DataType::Int32 => true,
+        DataType::Int64 => true,
+        DataType::UInt8 => true,
+        DataType::UInt16 => true,
+        DataType::UInt32 => true,
+        DataType::UInt64 => true,
+        DataType::Float32 => true,
+        DataType::Float64 => true,
+        DataType::Timestamp(time_unit, _) => match time_unit {
+            TimeUnit::Seconds => true,
+            TimeUnit::Milliseconds => true,
+            TimeUnit::Microseconds => true,
+            TimeUnit::Nanoseconds => true,
+        },
+        DataType::Utf8 => true,
+
+        DataType::Decimal128(_, _) => true,
+        DataType::Date => true,
+
+        DataType::FixedSizeBinary(_) => true,
+
+        DataType::List(_) => true,
+
+        DataType::FixedSizeList(_, _) => true,
+        DataType::Struct(fields) => fields.iter().all(|f| can_hash(&f.dtype)),
+        _ => false,
+    }
+}
```

Comment:
Huh, I wonder what the behavior right now is if you specify a join key that is not hashable. Additionally, should this be a method for `DataType` instead?